### PR TITLE
Sstoolkit 34

### PR DESCRIPTION
**Pull request by unserializable**
* Added command 'activate' for auth certificate activation (enabling) -- ``xrdsst cert activate``, depends on imported and registered auth certificate.

EDIT:

* Added ``cert download-csrs`` command

### CODE REVIEW

File: tests/unit/test_cert.py [hunk 1]
```diff
@@ -344,8 +344,28 @@ def test_cert_register_multiple_certs_for_auth_key(self):
...
 
                 out, err = self.capsys.readouterr()
-                assert out.count("Multiple registrable certificates for key") > 0
+                assert out.count("Multiple certificates to 'REGISTER' for key") > 0
 
                 with self.capsys.disabled():
                     sys.stdout.write(out)
                     sys.stderr.write(err)
+
+    def test_cert_activate(self):
```
Comment (bertvi) [inline tests/unit/test_cert.py:447]: Please also test failures/exceptions in addition to the happy path.

**Comment by bertvi** [comment]
Integration tests fail. Failure is related to PyLint errors, which should be fixed.

=============================================================================================================== FAILURES ===============================================================================================================
_____________________________________________________________________________________________ [pylint] tests/unit/test_base_controller.py ______________________________________________________________________________________________
W:129, 8: Access to a protected member _parser of a client class (protected-access)
W:130,22: Access to a protected member _parser of a client class (protected-access)
W:132, 8: Access to a protected member _pre_argument_parsing of a client class (protected-access)
R:127, 4: Method could be a function (no-self-use)
R:135, 4: Method could be a function (no-self-use)
___________________________________________________________________________________________________ [pylint] tests/unit/test_cert.py ___________________________________________________________________________________________________
C: 60, 0: Line too long (1049/160) (line-too-long)
C: 62, 0: Line too long (550/160) (line-too-long)
C:264,34: More than one statement on a single line (multiple-statements)
R:264,12: Method could be a function (no-self-use)
R:  1, 0: Similar lines in 6 files
==tests.unit.test_cert:132
==tests.unit.test_cert:147
==tests.unit.test_cert:162
==tests.unit.test_cert:178
==tests.unit.test_cert:193
==tests.unit.test_token:24
        available=True,
        id=0,
        logged_in=True,
        name='softToken-0',
        possible_actions=None,
        read_only=False,
        saved_to_configuration=True,
        serial_number=None,
        status=TokenStatus.OK,
        token_infos=[{'key': 'Type'}, {'value': 'Software'}],
        type=TokenType.SOFTWARE, (duplicate-code)
R:  1, 0: Similar lines in 2 files
==tests.unit.test_cert:210
==tests.unit.test_token:125
    ss_config = {
        'logging': [{'file': '/tmp/xrdsst_test_token_log', 'level': 'INFO'}],
        'api_key': [{'url': 'https://localhost:4000/api/v1/api-keys',
                     'key': 'private key',
                     'credentials': 'user:pass',
                     'roles': 'XROAD_SYSTEM_ADMINISTRATOR'}],
        'security_server':
            [{'name': 'ssX',
              'url': 'https://non.existing.url.blah:8999/api/v1', (duplicate-code)
R:  1, 0: Similar lines in 2 files
==tests.unit.test_cert:224
==tests.unit.test_token:136
              'owner_dn_country': 'FI',
              'owner_dn_org': 'UNSERE',
              'owner_member_class': 'VOG',
              'owner_member_code': '4321',
              'security_server_code': 'SS3',
              'software_token_id': '0',
              'software_token_pin': '1122'}]}
 (duplicate-code)
R:  1, 0: Similar lines in 2 files
==tests.unit.test_cert:34
==tests.unit.test_token:58
                        not_before=datetime(2020, 11, 5, 22, 24, 45, tzinfo=tzutc()),
                        public_key_algorithm='RSA',
                        rsa_public_key_exponent=65537,
                        rsa_public_key_modulus=214432,  # Not even a prime, but fine ATM
                        signature='30eb0bde826774031726',
                        signature_algorithm='SHA256withRSA',
                        serial=5, (duplicate-code)
__________________________________________________________________________________________________ [pylint] tests/unit/test_xrdsst.py __________________________________________________________________________________________________
C: 31, 0: Final newline missing (missing-final-newline)
E: 11,11: Class 'JustObject' has no 'OP_GRAPH' member (no-member)
E: 12,11: Class 'JustObject' has no 'OP_DEPENDENCY_LIST' member (no-member)
E: 13,11: Class 'JustObject' has no 'OP_GRAPH' member (no-member)
E: 13,54: Class 'JustObject' has no 'OP_DEPENDENCY_LIST' member (no-member)
C: 18,11: Comparison should be BaseController.default_sign_key_label(security_server_config) == 'ss-name1-default-sign-key' (misplaced-comparison-constant)
C: 22,11: Comparison should be BaseController.default_auth_key_label(security_server_config) == 'ss-name2-default-auth-key' (misplaced-comparison-constant)
C: 27,11: Comparison should be BaseController.security_server_address(security_server_config) == 'somehost' (misplaced-comparison-constant)
C: 29,11: Comparison should be BaseController.security_server_address(security_server_config) == 'otherhost' (misplaced-comparison-constant)
C: 31,11: Comparison should be BaseController.security_server_address(security_server_config) == '192.168.1.33' (misplaced-comparison-constant)

----------- coverage: platform linux, python 3.8.5-final-0 -----------
Name                              Stmts   Miss  Cover
-----------------------------------------------------
xrdsst/controllers/base.py          105     12    89%
xrdsst/controllers/cert.py          112      9    92%
xrdsst/controllers/init.py           58      3    95%
xrdsst/controllers/timestamp.py      95      9    91%
xrdsst/controllers/token.py         155     14    91%
-----------------------------------------------------
TOTAL                               525     47    91%
Coverage HTML written to dir coverage-report

======================================================================================================= short test summary info ========================================================================================================
FAILED tests/unit/test_base_controller.py::PYLINT
FAILED tests/unit/test_cert.py::PYLINT
FAILED tests/unit/test_xrdsst.py::PYLINT
=============================================================================================== 3 failed, 76 passed in 85.34s (0:01:25) ================================================================================================
make: *** [Makefile:35: test-all] Error 1

**Comment by bertvi** [comment]
Please also add integration tests for "cert import", "cert register" and "cert activate".

**Review by bertvi — CHANGES_REQUESTED** [review]

File: tests/unit/test_cert.py (hunk 1 above)
Comment (unserializable) [reply tests/unit/test_cert.py:447]: Any specific scenario do in mind here, most failures that I could see are tested already?

Comment (unserializable) [reply tests/unit/test_cert.py:447]: > Please also add integration tests for "cert import", "cert register" and "cert activate".

I would rather not tackle this until the Jenkins access is sorted out, running them locally is far too time consuming.

Comment (unserializable) [reply tests/unit/test_cert.py:447]: > Integration tests fail. Failure is related to PyLint errors, which should be fixed.
Most of these PyLint errors are there for a while and are quite absurd, do you think it makes any sense to add ignores for them?

As for integration tests failing -- linting just should not fail integration tests, they should be in separate targets (``make check`` is one of the conventional target names sometimes used for that).

Comment (bertvi) [reply tests/unit/test_cert.py:447]: > Any specific scenario do in mind here, most failures that I could see are tested already?

What if these throw an exception: 

        token_cert_api.activate_certificate(activatable_cert.certificate_details.hash) # responseless PUT
        cert_actions = token_cert_api.get_possible_actions_for_certificate(activatable_cert.certificate_details.hash)

Comment (unserializable) [reply tests/unit/test_cert.py:447]: There is no specific handler for these, so nothing specific to test.

**Comment by bertvi** [comment]
Please fix these Linter warnings:

tests/integration/__init__.py:1:0: R0801: Similar lines in 2 files
==xrdsst.controllers.cert:229
==xrdsst.controllers.token:210
def remote_get_token(ss_configuration, security_server):
    token_id = security_server['software_token_id']
    token_api = TokensApi(ApiClient(ss_configuration))
    token = token_api.get_token(token_id)
    return token


tests/integration/test_xrdsst.py:212:12: W0612: Unused variable 'result' (unused-variable)
tests/integration/test_xrdsst.py:219:12: W0612: Unused variable 'result' (unused-variable)
tests/integration/test_xrdsst.py:226:12: W0612: Unused variable 'result' (unused-variable)


tests/unit/test_cert.py:297:12: R1705: Unnecessary "elif" after "return" (no-else-return)
tests/unit/test_cert.py:296:32: W0613: Unused argument 'self' (unused-argument)
tests/unit/test_cert.py:296:38: W0613: Unused argument 'id' (unused-argument)
tests/unit/test_cert.py:296:0: W0613: Unused argument 'kwargs' (unused-argument)

xrdsst/controllers/cert.py:103:8: C0415: Import outside toplevel (cement.utils.fs) (import-outside-toplevel)
xrdsst/controllers/cert.py:169:20: C0415: Import outside toplevel (cement.utils.fs) (import-outside-toplevel)

**Comment by unserializable** [comment]
> tests/unit/test_cert.py:296:38: W0613: Unused argument 'id' (unused-argument)
> tests/unit/test_cert.py:296:0: W0613: Unused argument 'kwargs' (unused-argument)

Method with certain signature is mocked, first arguments are positional, besides chopping off existing method signatures is often a code smell even in dynamic languages.

**Pull request merged**